Add support for GMSA CredentialSpecs from Swarmkit configs

dperny opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Adds the ability to specify a swarmkit Config as a CredentialSpec, as an alternative to the existing registry and file options. Adds a new CredentialSpec URI specifier: `config://`. This is used to specify the ID of a swarmkit config to be used as the CredentialSpec.

**- How I did it**

Altered swarmkit to support distributing a config that is not mounted into a container directly. Then, altered the `oci_windows.go` code to get that config from the dependency store, and use the data as the GMSA config.

**- How to verify it**

This will need to be manually tested, as I do not have access to a Windows environment on which to test.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Add support for using swarmkit Configs as GMSA CredentialSpecs.

dperny committed:
Add CredentialSpec from configs support

Signed-off-by: Drew Erny <drew.erny@docker.com>

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "gmsa-support" git@github.com:dperny/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842357812584
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

dperny commented:
Interestingly, this requires no CLI or API changes, because CredentialSpecs are specified using that weird URI format.

cpuguy83 commented:
I assume this has that change for async delete in it? We need to resolve that before we can re-vendor swarmkit.

dperny commented:
@cpuguy83 yes, sorry, I should have mentioned that. This is blocked on the async delete changes.

dperny commented:
Ok, the async delete change has been reverted, and I have revendored swarmkit, so this commit no longer includes that change. This is fully ready for review.

thaJeztah committed:
Add support for using Configs as CredentialSpecs in services

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
oh, wha, missed this one; guess this will have to be combined with https://github.com/moby/moby/pull/38672 ?

dperny commented:
@thaJeztah yes, affirmative.

dperny commented:
This PR is updated

* Removed the vendoring commit, as swarmkit is already vendored to the correct version
* Added @thaJeztah's commit from #38672 to this branch
* Rebased

thaJeztah commented on daemon/oci_windows.go:
```diff
@@ -288,6 +288,23 @@ func (daemon *Daemon) createSpecWindowsFields(c *container.Container, s *specs.S
+				}
+				csConfig, err := c.DependencyStore.Configs().Get(csValue)
+				if err != nil {
+					return fmt.Errorf("error getting value from config store: %v", err)
```
This should probably use `errors.Wrap` to preserve the original error; something like;

```go
return errors.Wrap(err, "failed to get value from config store")
```

dperny replied:
Good catch.

thaJeztah commented on daemon/oci_windows.go:
```diff
@@ -288,6 +288,23 @@ func (daemon *Daemon) createSpecWindowsFields(c *container.Container, s *specs.S
+
+				// if the container does not have a DependencyStore, then we
+				// return an error
+				if c.DependencyStore == nil {
```
Wondering (reading from my phone so have limited sight in the whole code path) if we can catch this situation during _create_ of the container instead of here (as this is (I think) only ran at the last point, at container start)

eg somewhere in this code path (would have to check if that's the right place though) https://github.com/moby/moby/blob/e7a9a7cdbc8fc87cb2bf07b770bd8bfe52ab0959/daemon/create.go#L25

/cc @cpuguy83

thaJeztah reviewed (changes requested):
Just some quick comments from my phone 😅

dperny commented:
@thaJeztah are you going to veto this if I push any code-review-stage changes to a new commit, instead of reordering the two commits i have or altering your commit (which is at the head of the branch)?

cpuguy83 commented:
Is it not odd that we are pulling in swarm data at the oci config level?

dperny commented:
It is a bit odd, but where else would we do it? This is the part of the code where we get the actual CredentialSpec data.

cpuguy83 commented:
Can we create a credential file from the config and use that in the host config?
Potentially could be done after create if we allow the swarm agent to update the container config after create.

dperny commented:
Seems like an awful lot of runaround, especially if it involves a temporary file. The DependencyStore is part of the `container.Container` type, so why not use it? Plus, it's indirected through an interface, so it's plenty decoupled from swarm.

thaJeztah commented:
> are you going to veto this if I push any code-review-stage changes to a new commit, instead of reordering the two commits i have or altering your commit (which is at the head of the branch)?

if you do a "rebase -i" you can edit the previous commit if needed.

I'm on PTO, and not near a computer until later tonight or otherwise tomorrow, so reviewing "by proxy" a bit;  hence I pinged Brian to help 🤗

> On 5 Feb 2019, at 18:08, Drew Erny <notifications@github.com> wrote:
> 
> @thaJeztah are you going to veto this if I push any code-review-stage changes to a new commit, instead of reordering the two commits i have or altering your commit (which is at the head of the branch)?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

dperny commented:
Ugh I realize I've bungled some things up.

In Swarmkit, the Config referenced in a CredentialSpec needs to be included in the ContainerSpec's ConfigReferences, with a `RuntimeTarget` instead of a `FileTarget`. This tells Swarmkit, "This config is going to be used somewhere, but not as a file in the container"

Then, the CredentialSpec is set to use a Config and references the ConfigID. This tells the daemon, "when you want a CredentialSpec, use this config".

The reason for this being in two places is that the ConfigReferences tell you what configs you're using, and the CredentialSpec tells you what CredentialSpec you're using. This means any existing logic for Configs is contained totally to checking ConfigReferences, and does not need to shop around for the other places Configs could be hiding. The biggest benefit for this is in the validation logic, and in the dispatcher logic.

My biggest mistake so far is that I forgot to write a validation check that ensures the config referenced as a CredentialSpec is actually included in the ConfigReferences, so nothing stops you from putting a Config in the CredentialSpec that can't be used. 

I've been assuming we'll take the same pattern with the Docker API, but I neglected to take into account the fact that the `RuntimeTarget` object is empty -- it doesn't need any data (yet). I'm unsure if JSON would allow us to specify an empty object that is semantically different from a nil object...

The ultimate intention at the CLI level was for the user to simply pass `--credential-spec="configid"`, and then have the CLI go and get the Config and add it to both the ConfigReferences and the CredentialSpec field.

dperny committed:
Use Runtime target

The Swarmkit api specifies a target for configs called called "Runtime"
which indicates that the config is not mounted into the container but
has some other use. This commit updates the Docker api to reflect this.

Signed-off-by: Drew Erny <drew.erny@docker.com>

dperny commented:
Updated the API to reflect the `Runtime` target.

dperny commented on api/types/swarm/config.go:
```diff
@@ -27,9 +27,14 @@ type ConfigReferenceFileTarget struct {
 // ConfigReference is a reference to a config in swarm
 type ConfigReference struct {
-	File       *ConfigReferenceFileTarget
+	File       *ConfigReferenceFileTarget    `json:",omitempty"`
```
Does adding `json:",omitempty"` alter the API in an incompatible way?

thaJeztah commented:
Looks like there's a gofmt failure;

```
20:45:09 daemon/cluster/convert/service_test.go:1::warning: file is not gofmted with -s (gofmt)
20:45:10 Build step 'Execute shell' marked build as failure
```

thaJeztah commented on api/server/router/swarm/cluster_routes.go:
```diff
@@ -270,6 +280,11 @@ func (sr *swarmRouter) updateService(ctx context.Context, w http.ResponseWriter,
+				if service.TaskTemplate.ContainerSpec.Privileges != nil && service.TaskTemplate.ContainerSpec.Privileges.CredentialSpec != nil {
+					// Support for setting credential-spec through configs was added in API 1.40
+					service.TaskTemplate.ContainerSpec.Privileges.CredentialSpec.Config = ""
+				}
```
I guess we need to strip the `Runtime` property here as well (as is done in `createService`)?
```suggestion
				}
				for _, config := range service.TaskTemplate.ContainerSpec.Configs {
					// support for the Runtime target was added in API 1.40
					config.Runtime = nil
				}
```

thaJeztah commented on api/types/swarm/config.go:
```diff
@@ -27,9 +27,14 @@ type ConfigReferenceFileTarget struct {
 type ConfigReference struct {
-	File       *ConfigReferenceFileTarget
+	File       *ConfigReferenceFileTarget    `json:",omitempty"`
+	Runtime    *ConfigReferenceRuntimeTarget `json:",omitempty"`
```
Can you add this field to the API history documentation as well? https://github.com/moby/moby/blob/master/docs/api/version-history.md#v140-api-changes

thaJeztah commented on daemon/cluster/convert/container.go:
```diff
@@ -191,7 +185,15 @@ func configReferencesToGRPC(sr []*types.ConfigReference) []*swarmapi.ConfigRefer
+		// non-existance of s.File to show that this isn't a File. Instead, the
+		// existence of the Runtime field marks this as a Runtime reference
+		// instead of a File reference.
+		if s.Runtime != nil {
```
We should probably also add a validation step before we reach this, and generate an `errdef.InvalidRequest` if both `File` and `Runtime` are set, as they are (at least currently) not allowed to be both set;

```go
func configReferencesToGRPC(sr []*types.ConfigReference) ([]*swarmapi.ConfigReference, error) {
	refs := make([]*swarmapi.ConfigReference, 0, len(sr))
	for _, s := range sr {
		ref := &swarmapi.ConfigReference{
			ConfigID:   s.ConfigID,
			ConfigName: s.ConfigName,
		}
		switch {
		case s.Runtime != nil && s.File != nil:
			return nil, errors.New("cannot specify both File and Runtime")
		case s.Runtime != nil:
			// Runtime target was added in API v1.40 and takes precedence over
			// File target. However, File and Runtime targets are mutually exclusive,
			// so we should never have both.
			ref.Target = &swarmapi.ConfigReference_Runtime{
				Runtime: &swarmapi.RuntimeTarget{},
			}
		case s.File != nil:
			ref.Target = &swarmapi.ConfigReference_File{
				File: &swarmapi.FileTarget{
					Name: s.File.Name,
					UID:  s.File.UID,
					GID:  s.File.GID,
					Mode: s.File.Mode,
				},
			}
		}

		refs = append(refs, ref)
	}

	return refs, nil
}
```

I'll post a full diff at the end

thaJeztah commented on docs/api/version-history.md:
```diff
@@ -29,6 +29,8 @@ keywords: "API, Docker, rcli, REST, documentation"
 * `POST /services/create` now accepts `Sysctls` as part of the `ContainerSpec`.
 * `POST /services/{id}/update` now accepts `Sysctls` as part of the `ContainerSpec`.
+* `POST /services/create` now accepts `Config` as part of `ContainerSpec.Privileges.CredentialSpec`.
+* `POST /services/{id}/update` now accepts `Config` as part of `ContainerSpec.Privileges.CredentialSpec`.
```
```suggestion
* `POST /services/{id}/update` now accepts `Config` as part of `ContainerSpec.Privileges.CredentialSpec`.
* `POST /services/create` and `POST /services/{id}/update` now accept a `Runtime`
  field in `ContainerSpec.Configs[].Config`.
```

thaJeztah reviewed (changes requested):
Left comments inline, but was testing my proposed changes locally, so I attached a patch with all the changes I proposed (added `.txt` extension to make GitHub happy)

[pr-38632.patch.txt](https://github.com/moby/moby/files/2845302/pr-38632.patch.txt)

dperny commented on api/swagger.yaml:
```diff
@@ -2756,6 +2768,9 @@ definitions:
             items:
               type: "object"
               properties:
+                Runtime:
```
The way it's done here is actually closer to the way it's implemented in the Swarmkit api

dperny commented on api/server/router/swarm/cluster_routes.go:
```diff
@@ -218,13 +218,23 @@ func (sr *swarmRouter) createService(ctx context.Context, w http.ResponseWriter,
+				}
+				for _, config := range service.TaskTemplate.ContainerSpec.Configs {
+					// support for the Runtime target was added in API 1.40
+					config.Runtime = nil
```
A Config with no Target is invalid. But I believe that having a Runtime-type Config is important because it affirms the intended purpose of the config.

A Config with both a Runtime and a File should be invalid, but I believe that the current implementation would just favor Runtime and omit the File portion entirely. I'm unsure where the validation check should be.

Runtime isn't the best name but that's what I called it in Swarmkit and that's what we merged, so I'm committing to it.

cpuguy83 commented:
> Seems like an awful lot of runaround, especially if it involves a temporary file

The problem is two fold:

1. Exposing swarm objects at this level is just not great... tbh the whole thing is not great and this adds more spaghetti to the mix.
2. This exposes `config://` as an API on the container runtime side.

I would really like to find a way to attach a config to a container w/o exposing at this level.

dperny commented:
@thaJeztah should have your comments all addressed.

thaJeztah reviewed (approved):
thanks! last changes LGTM, but I see @cpuguy83 has concerns on the implementation

dperny commented:
I don't think it's inappropriate to expose at this level though. We're already exposing the container object, which has access to the dependency store. I'd argue that this access to the dependency store is here for precisely this reason -- to get secret or config data whenever it is needed.

cpuguy83 commented:
The dependency store was added as a hack instead of adding an internal api to attach objects to the container, which is part of the problem.
At this level we should not be dealing with swarm related things at all.

dperny commented:
I'm still unsure of how else to go about it without passing through a temporary file (or temporary registry entry, I guess) or re-writing a bunch of method signatures, both of which seem like even worse options.

dperny commented:
I don't have a good enough understanding of the process of writing out and mounting secrets and configs in the windows environment. I theoretically Secrets and Configs should be mounted somewhere on the host that isn't readable by the container, and then linked to a location that is readable in the container? If so, how would I go about only writing the Config to the common location, and not to the container itself?

dperny commented:
Copying the CredentialSpec onto the host and reusing the `file://` or `registry://` targets creates its own problems with mixing concerns. Currently, as far as I can tell, the paths for Credential Specs, for both `file://` and `registry://`, are defined as private constants in `daemon/oci_windows.go`, but the swarmkit cluster-related code is all in the `daemon/cluster/` package.

I think it may actually be the case that adding a `config://` path to the CredentialSpec security options may be the best way to go about this.

1. `oci_windows.go` is the point where we translate to the container runtime spec, and is still totally within the context of the daemon, so this isn't a super inappropriate place to reference the DependencyStore
2. The Swarmkit bits are already depended on in this code and on swarmkit features in general. The `createSpec` method calls `daemon.setupSecretDir` and `daemon.setupConfigDir`, both of which use the DependencyStore.

Further, there is one major headache with copying the CredSpecs onto the host: cleanup. If the node crashes, we have to make sure that when it comes back, all the no-longer-in-use CredentialSpecs are cleaned up.

dperny commented:
Updated to address an issue I noticed, where error messages would be logged for Runtime-type configs when doing `daemon.setupConfigDir`.

dperny commented:
Updated to change the error message when the container does not have a DependencyStore. WIth this, we don't return any errors that create any impression that `config://` prefix is in any way part of the API.

cpuguy83 commented on daemon/container_operations_windows.go:
```diff
@@ -44,7 +44,14 @@ func (daemon *Daemon) setupConfigDir(c *container.Container) (setupErr error) {
+			// Runtime configs are not mounted into the container, but they're
+			// a valid type of config so we should not error when we encounter
+			// one.
+			if configRef.Runtime == nil {
```
Maybe move this up  so it's `if configRef.File == nil && configRef.Runtime == nil`?

dperny replied:
Nah, because in either case you're going to `continue`, so you'd end up writing

```go
if configRef.File == nil && configRef.Runtime == nil {
    logrus.Error("config target type is not a file or runtime target")
    continue
}
if configRef.Runtime != nil {
    continue
}
```

cpuguy83 replied:
Ah I see.

cpuguy83 commented on daemon/oci_windows.go:
```diff
@@ -288,6 +288,28 @@ func (daemon *Daemon) createSpecWindowsFields(c *container.Container, s *specs.S
+				// impression that `config://` is a valid API, return the same
+				// error as if you'd passed any other random word.
+				if c.DependencyStore == nil {
+					return fmt.Errorf("invalid credential spec security option - value must be prefixed file:// or registry:// followed by a value")
```
`errdefs.InvalidParameter(errors.Errorf(...))`

dperny replied:
I can do that, but this function is littered with `return fmt.Errorf` calls. It is called by `createSpec`, which is in turn called in `(*Daemon).containerStart`, where the error result is wrapped (in `daemon/start.go`, on line 154, with `errdefs.System`:

https://github.com/moby/moby/blob/740349757396d8f1ad573d4b78148baca9c979aa/daemon/start.go#L154-L157

cpuguy83 replied:
🤦‍♂️ That error conversion is incorrect.

cpuguy83 commented on daemon/oci_windows.go:
```diff
@@ -288,6 +288,28 @@ func (daemon *Daemon) createSpecWindowsFields(c *container.Container, s *specs.S
+				// after this point, we can return regular swarmkit-relevant
+				// errors, because we'll know this container is managed.
+				if csValue == "" {
+					return fmt.Errorf("no value supplied for config:// credential spec security option")
```
`errdefs.InvalidParameter`

cpuguy83 commented on api/server/router/swarm/helpers.go:
```diff
@@ -64,3 +66,32 @@ func (sr *swarmRouter) swarmLogs(ctx context.Context, w io.Writer, r *http.Reque
+// adjustForAPIVersion takes a version and service spec and removes fields to
+// make the spec compatible with the specified version.
+func adjustForAPIVersion(cliVersion string, service *swarm.ServiceSpec) {
+	if cliVersion != "" {
```
```go
if cliVersion == "" { 
    return
}
```

dperny replied:
Done

cpuguy83 commented on daemon/container_operations_unix.go:
```diff
@@ -230,7 +230,14 @@ func (daemon *Daemon) setupSecretDir(c *container.Container) (setupErr error) {
+			// Runtime configs are not mounted into the container, but they're
+			// a valid type of config so we should not error when we encounter
+			// one.
+			if ref.Runtime == nil {
```
Maybe move this up so it's `if configRef.File == nil && configRef.Runtime == nil`?

wk8 commented on api/server/router/swarm/cluster_routes.go:
```diff
@@ -213,19 +213,7 @@ func (sr *swarmRouter) createService(ctx context.Context, w http.ResponseWriter,
-				service.TaskTemplate.Placement.MaxReplicas = 0
-			}
-		}
+		adjustForAPIVersion(cliVersion, &service)
```
Actually, could be nice to add the `queryRegistry` toggling in there too, eg by passing a `*bool` as 3rd argument (since returning a bool feels like a weird API)?

dperny replied:
I considered this, but decided against it.

wk8 commented on daemon/cluster/convert/container.go:
```diff
@@ -10,6 +9,7 @@ import (
 	types "github.com/docker/docker/api/types/swarm"
 	swarmapi "github.com/docker/swarmkit/api"
 	gogotypes "github.com/gogo/protobuf/types"
+	"github.com/pkg/errors"
```
Nice, I'll remember this one. :)

wk8 commented on daemon/oci_windows.go:
```diff
@@ -288,6 +288,28 @@ func (daemon *Daemon) createSpecWindowsFields(c *container.Container, s *specs.S
+				// if the container does not have a DependencyStore, then it
+				// isn't swarmkit managed. In order to avoid creating any
+				// impression that `config://` is a valid API, return the same
+				// error as if you'd passed any other random word.
```
Maybe you could make that error a package variable then? That would ensure it stays the same.

dperny replied:
I'm not too concerned because it's evident that all of the errors in this function need to be overhauled anyway.

cpuguy83 reviewed (approved):
LGTM

dperny commented:
bless 🙌
